fenicsadapter/adapter_core.py

```diff
@@ -0,0 +1,194 @@
+    rank evaluation.
+    """
+    if input_function.value_rank() == 0:  # scalar-valued functions have rank 0 is FEniCS
+        return FunctionType.SCALAR
+    elif input_function.value_rank() == 1:  # vector-valued functions have rank 1 in FEniCS
+        return FunctionType.VECTOR
+    else:
+        raise Exception("Error determining function type")
+
+
+def filter_point_sources(point_sources, filter_out):
+    """
+    Filter dictionary of PointSources (point_sources) with respect to a given domain (filter_out). If a PointSource
+    is applied at a point inside of the given domain (filter_out), this PointSource will be removed from dictionary.
+    :param point_sources: dictionary containing coordinates and associated PointSources;
+      {(point_x, point_y): PointSource, ...}
+    :param filter_out: defines the domain where PointSources should be filtered out
+    :return: A dictionary with the filtered PointSources
+    """
+
+    filtered_point_sources = dict()
+
+    for point in point_sources.keys():
+        # Filter double boundary points to avoid instabilities and create PointSource
+        if filter_out.inside(point, 1):
+            print("Found a double-boundary point at {location}.".format(location=point))
+        else:
+            filtered_point_sources[point] = point_sources[point]
+
+    return filtered_point_sources
+
+
+def convert_fenics_to_precice(data, sample_points):
+    """Converts FEniCS data of type dolfin.Function into Numpy array for all x and y coordinates on the boundary.
+
+    :param data: FEniCS boundary function
+    :param sample_points: Vertices
+    :raise Exception: if type of data cannot be handled
+    :return: array of FEniCS function values at each point on the boundary
+    """
+    if type(data) is dolfin.Function:
+        x_all, y_all = sample_points[:, 0], sample_points[:, 1]
+        return np.array([data(x, y) for x, y in zip(x_all, y_all)])
+    else:
+        raise Exception("Cannot handle data type %s" % type(data))
+
+
+def get_coupling_boundary_vertices(mesh_fenics, coupling_subdomain, fenics_dimensions, dimensions):
+    """Extracts vertices which lie on the boundary.
+    :return: stack of vertices
+    """
+    n = 0
+    fenics_vertices = []
+    vertices_x = []
+    vertices_y = []
+    if dimensions == 3:
+        vertices_z = []
+
+    if not issubclass(type(coupling_subdomain), SubDomain):
+        raise Exception("no correct coupling interface defined!")
+
+    for v in dolfin.vertices(mesh_fenics):
+        if coupling_subdomain.inside(v.point(), True):
+            n += 1
+            fenics_vertices.append(v)
+            vertices_x.append(v.x(0))
+            if dimensions == 2:
+                vertices_y.append(v.x(1))
+            elif fenics_dimensions == 2 and dimensions == 3:
+                vertices_y.append(v.x(1))
+                vertices_z.append(0)
+            else:
+                raise Exception("Dimensions do not match!")
```

Comment:
Dimensions of what do not match? Could we tell the user in the message what is going wrong more specifically or is it clear from context?